pkg/gitrepository/manager_test.go

```diff
@@ -0,0 +1,193 @@
+package gitrepository_test
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/fluxcd/pkg/apis/meta"
+	sourcev1 "github.com/fluxcd/source-controller/api/v1beta1"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	profilesv1 "github.com/weaveworks/profiles/api/v1alpha1"
+	"github.com/weaveworks/profiles/pkg/gitrepository"
+	"github.com/weaveworks/profiles/pkg/gitrepository/fakes"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var _ = Describe("Gitrepository", func() {
+	var (
+		manager   *gitrepository.Manager
+		kClient   *fakes.FakeKubernetes
+		callCount int
+		repo      = profilesv1.Repository{
+			URL: "github.com/example/repo",
+			SecretRef: &meta.LocalObjectReference{
+				Name: "my-secret",
+			},
+		}
+	)
+
+	BeforeEach(func() {
+		callCount = 0
+		kClient = new(fakes.FakeKubernetes)
+		manager = gitrepository.NewManager(context.TODO(), "profiles-system", kClient, time.Second, time.Millisecond)
+	})
+
+	When("the gitrepositorys create successfully", func() {
+		BeforeEach(func() {
+			kClient.GetStub = func(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+				callCount++
+				if callCount == 3 {
+					Expect(obj.(*sourcev1.GitRepository).Name).To(Equal("repo-v0.1.0"))
+					obj.(*sourcev1.GitRepository).Status = sourcev1.GitRepositoryStatus{
+						URL: "url1",
+					}
+				}
+				if callCount == 4 {
+					Expect(obj.(*sourcev1.GitRepository).Name).To(Equal("repo-foo-v1.0.0"))
+					obj.(*sourcev1.GitRepository).Status = sourcev1.GitRepositoryStatus{
+						URL: "url2",
+					}
+				}
+				return nil
+			}
+		})
+
+		It("creates the gitrepository resources and waits for them to be ready", func() {
+			resources, err := manager.CreateAndWaitForResources(repo, []gitrepository.Instance{
+				{
+					Tag:  "v0.1.0",
+					Path: "profile.yaml",
+				},
+				{
+					Tag:  "foo/v1.0.0",
```

Comment:
Add invalid tags here, like, `foo_whatever` which shouldn't break things after the tag name is sanitised.